File: hackerrank/problem solving/data structures/1 - arrays/2 - 2d array - ds/2d_array_ds.rs
// https://www.hackerrank.com/challenges/2d-array/problem?isFullScreen=true

use std::cmp;
use text_io::read;

fn main() {
    let matrix: Vec<Vec<i32>> = read_matrix();
    println!("{}", max_hourglass_sum(&matrix));
}

fn read_matrix() -> Vec<Vec<i32>> {
    const N: usize = 6;
    let matrix: Vec<Vec<i32>> = (0..N).map(|_| (0..N).map(|_| read!()).collect()).collect();
    matrix
}

// T: O(1)
// S: O(1) extra space
fn max_hourglass_sum(matrix: &[Vec<i32>]) -> i32 {
    let mut max_sum: i32 = i32::MIN;
    const ROWS_COLS: usize = 4;

    for i in 1..=ROWS_COLS {
        for j in 1..=ROWS_COLS {
            let current_sum: i32 = hourglass_sum(matrix, i, j);
            max_sum = cmp::max(max_sum, current_sum);
        }
    }

    max_sum
}

fn hourglass_sum(matrix: &[Vec<i32>], i: usize, j: usize) -> i32 {
    let subrow1_first_index: usize = i - 1;
    let subrow3_first_index: usize = i + 1;
    let first_col_index: usize = j - 1;
    let end: usize = first_col_index + 3;

    let subrow1_sum: i32 = matrix[subrow1_first_index][first_col_index..end]
        .iter()
        .sum();
    let subrow3_sum: i32 = matrix[subrow3_first_index][first_col_index..end]
        .iter()
        .sum();

    subrow1_sum + matrix[i][j] + subrow3_sum
}

File: hackerrank/problem solving/data structures/1 - arrays/2 - 2d array - ds/2d_array_ds.cpp
// https://www.hackerrank.com/challenges/2d-array/problem?isFullScreen=true

#include <algorithm>
#include <iostream>
#include <numeric>
#include <valarray>

using namespace std;

int max_hourglass_sum(const valarray<int>& matrix);
int hourglass_sum(const valarray<int>& matrix, const int index);

constexpr int stride {6};

int main()
{
    constexpr int n {6};
    valarray<int> matrix(n * n);
    ranges::generate(matrix, [] {int x; cin >> x; return x;});
    cout << max_hourglass_sum(matrix);

    return 0;
}

    int max_hourglass_sum(const valarray<int>& matrix)
    {
        int max_sum {numeric_limits<int>::min()};
        constexpr int middle_hourglass_first_index {7};
        constexpr int n_sub_rows_cols {4};
        constexpr int n {stride * n_sub_rows_cols + 1};

        for (int i {middle_hourglass_first_index}; i <= n; i += stride)
            for (int j {i}, cols {4}, n2 {j + cols}; j < n2; ++j) {
                int current_sum {hourglass_sum(matrix, j)};
                max_sum = max(max_sum, current_sum);
            }

        return max_sum;
    }

        int hourglass_sum(const valarray<int>& matrix, const int index)
        {
            int subrow_1_first_index {index - stride - 1};
            int subrow_3_first_index {index + stride - 1};
            constexpr int n {3};
            constexpr int stride {1};

            auto subrow_1 {slice(subrow_1_first_index, n, stride)};
            auto subrow_3 {slice(subrow_3_first_index, n, stride)};

            return matrix[subrow_1].sum() + matrix[index] + matrix[subrow_3].sum();
        }
